Check for bool conversion on Expressions

IgnaceBleukx opened the pull request:
I realized that we can catch some beginner mistakes where a user uses an CPMpy expression as bool.

For example:
```python
if any(cp.boolvar(shape=3)): ...
```
Which should actually raise an error or at least a warning, as it probably not what the user intended...

Now, this kind of opened pandora's box with a bunch of our tests failing because we did exactly this rookie mistake...
I fixed most tests, but there is one issue remaining still, where checking if two sets are equal, will actually also call the `__bool__` internally if the hashes collide (which is way the pytest test fails)

I'm not sure how to fix that currently, so happy to discuss. 
If we don't find a fix, we can maybe raise a warning instead? Or if we really want to, we can check in the stacktrace if the method was called from `set` (if it includes it of course...) I think @tias you looked into this in the context of CSE as well?

Dimosts commented:
This is a bit hacky, but will not add overhead, as it will be called only when `__bool__` is called in a comparison object.

And it is also quite clear and not cryptic, with a comment it would be fine.

As this may take some time to converge, on how to handle this, I suggest that the changes on the tests go to a separate pull request to be merged soon!! As there are many tests wrong currently.

IgnaceBleukx commented:
So I plugged the rabbit-hole and fixed all of the tests.
There really were are A LOT of issues uncovered by this simple sanity check.

Added some extra tests for checking of collections of expressions.
One test that did not make any sense to me is the inequality-chaining one, it basically tested if `True == True`, so removed it from this PR.

If this one is merged with the `__is_bool__` check in the `Comparison` object, I believe the CSE implementation will become much much much easier too.

ThomSerg reviewed (changes requested):
Indeed must have felt like opening a pandora's box. Aside from the comments, looks good to go.

IgnaceBleukx commented on cpmpy/expressions/globalconstraints.py:
```diff
@@ -515,7 +515,7 @@ def decompose(self):
         defining = []
         #if expr is not a var
-        if not isinstance(expr,_IntVarImpl):
+        if not isinstance(expr,Expression):
```
There was a test failing because of this (previously unnoticed as the test itself was using `if expr...`)

IgnaceBleukx replied:
Adapted to use `utils.get_bounds` instead

IgnaceBleukx commented:
Should be good to go, please merge if it's fine for you @ThomSerg : )

tias commented:
This is amazingly clean and clever, both wrt user experience, wrt fixing inaccurate tests, and wrt handling CSE without custom dicts...
